File: conformance/tests/escrow/escrow.runar.zig
const runar = @import("runar");

pub const Escrow = struct {
    pub const Contract = runar.SmartContract;

    buyer: runar.PubKey,
    seller: runar.PubKey,
    arbiter: runar.PubKey,

    pub fn init(buyer: runar.PubKey, seller: runar.PubKey, arbiter: runar.PubKey) Escrow {
        return .{
            .buyer = buyer,
            .seller = seller,
            .arbiter = arbiter,
        };
    }

    pub fn release(self: *const Escrow, sellerSig: runar.Sig, arbiterSig: runar.Sig) void {
        runar.assert(runar.checkSig(sellerSig, self.seller));
        runar.assert(runar.checkSig(arbiterSig, self.arbiter));
    }

    pub fn refund(self: *const Escrow, buyerSig: runar.Sig, arbiterSig: runar.Sig) void {
        runar.assert(runar.checkSig(buyerSig, self.buyer));
        runar.assert(runar.checkSig(arbiterSig, self.arbiter));
    }
};

File: conformance/tests/escrow/escrow.runar.go
//go:build ignore

package contract

import "runar"

type Escrow struct {
	runar.SmartContract
	Buyer   runar.PubKey `runar:"readonly"`
	Seller  runar.PubKey `runar:"readonly"`
	Arbiter runar.PubKey `runar:"readonly"`
}

func (c *Escrow) Release(sellerSig runar.Sig, arbiterSig runar.Sig) {
	runar.Assert(runar.CheckSig(sellerSig, c.Seller))
	runar.Assert(runar.CheckSig(arbiterSig, c.Arbiter))
}

func (c *Escrow) Refund(buyerSig runar.Sig, arbiterSig runar.Sig) {
	runar.Assert(runar.CheckSig(buyerSig, c.Buyer))
	runar.Assert(runar.CheckSig(arbiterSig, c.Arbiter))
}
